File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/models/Valence.kt
package app.pbbls.android.features.path.models

/**
 * Size axis of a valence — drives outline asset choice and render sizing.
 * [key] is the lowercase asset-name segment (`outline_<size>_<polarity>`).
 */
enum class ValenceSizeGroup(
    val key: String,
) {
    SMALL("small"),
    MEDIUM("medium"),
    LARGE("large"),
    ;

    companion object {
        /**
         * Top to bottom, the order the valence roll stacks sizes in — a big
         * event sits above a small one, so reaching the smaller ones means
         * scrolling down the ladder. Deliberately *not* [entries], which runs
         * small → large.
         */
        val ladder: List<ValenceSizeGroup> = listOf(LARGE, MEDIUM, SMALL)
    }
}

/** Polarity axis of a valence. [key] as in [ValenceSizeGroup.key]. */
enum class ValencePolarity(
    val key: String,
) {
    LOWLIGHT("lowlight"),
    NEUTRAL("neutral"),
    HIGHLIGHT("highlight"),
}

/**
 * The nine `(positiveness, intensity)` combinations — mirrors iOS
 * `Valence.swift`. Read-only in this milestone: only the derivation from DB
 * values is needed; the picker labels/images stay iOS-side until the create
 * flow ports.
 */
enum class Valence(
    val positiveness: Int,
    val intensity: Int,
) {
    LOWLIGHT_SMALL(-1, 1),
    LOWLIGHT_MEDIUM(-1, 2),
    LOWLIGHT_LARGE(-1, 3),
    NEUTRAL_SMALL(0, 1),
    NEUTRAL_MEDIUM(0, 2),
    NEUTRAL_LARGE(0, 3),
    HIGHLIGHT_SMALL(1, 1),
    HIGHLIGHT_MEDIUM(1, 2),
    HIGHLIGHT_LARGE(1, 3),
    ;

    val sizeGroup: ValenceSizeGroup
        get() =
            when (intensity) {
                1 -> ValenceSizeGroup.SMALL
                2 -> ValenceSizeGroup.MEDIUM
                else -> ValenceSizeGroup.LARGE
            }

    val polarity: ValencePolarity
        get() =
            when (positiveness) {
                -1 -> ValencePolarity.LOWLIGHT
                0 -> ValencePolarity.NEUTRAL
                else -> ValencePolarity.HIGHLIGHT
            }

    companion object {
        /**
         * Mirrors `Pebble.valence` on iOS: DB CHECK constraints guarantee the
         * pair, so an out-of-range value is decode drift — fall back to
         * [NEUTRAL_MEDIUM]. Deliberately log-free (pure JVM-tested code;
         * `android.util.Log` throws off-device) — `PathService` logs any
         * out-of-range pair after fetch.
         */
        fun fromOrDefault(
            positiveness: Int,
            intensity: Int,
        ): Valence =
            entries.firstOrNull { it.positiveness == positiveness && it.intensity == intensity }
                ?: NEUTRAL_MEDIUM

        /**
         * Null-returning variant for drafts (M47), where the pair may be absent
         * entirely. Deliberately does NOT fall back to [NEUTRAL_MEDIUM]: a draft
         * with no valence picked yet must hydrate the picker as unset rather than
         * claim a value the user never chose.
         */
        fun orNull(
            positiveness: Int?,
            intensity: Int?,
        ): Valence? {
            if (positiveness == null || intensity == null) return null
            return entries.firstOrNull { it.positiveness == positiveness && it.intensity == intensity }
        }
    }
}

/*
 * The roll — step geometry for the two-axis valence picker, porting the same
 * extension on iOS `Valence.swift`. Polarity runs left → right, size runs top
 * → bottom (`ValenceSizeGroup.ladder`).
 *
 * Pure index arithmetic, kept off the view so the roll's behaviour is asserted
 * without a gesture (`ValenceRollTest`).
 */

/**
 * The one case at a given cell. Total by construction: the 3×3 grid is
 * covered, which `ValenceRollTest` pins down.
 */
fun valenceAt(
    polarity: ValencePolarity,
    size: ValenceSizeGroup,
): Valence = Valence.entries.first { it.polarity == polarity && it.sizeGroup == size }

/**
 * The valence at the given indices, clamped to the grid — the roll stops at
 * the edges rather than wrapping, so a hard swipe cannot loop the user past
 * the end and back to where they started.
 */
fun valenceAt(
    polarityIndex: Int,
    sizeIndex: Int,
): Valence =
    valenceAt(
        polarity = ValencePolarity.entries[polarityIndex.coerceIn(0, ValencePolarity.entries.lastIndex)],
        size = ValenceSizeGroup.ladder[sizeIndex.coerceIn(0, ValenceSizeGroup.ladder.lastIndex)],
    )

/** Position on the polarity axis, in [ValencePolarity.entries] order. */
val Valence.polarityIndex: Int
    get() = ValencePolarity.entries.indexOf(polarity)

/** Position on the size axis, in the roll's top-to-bottom [ValenceSizeGroup.ladder] order. */
val Valence.sizeIndex: Int
    get() = ValenceSizeGroup.ladder.indexOf(sizeGroup)

/**
 * The polarity one step to the left, null at the end. Drives the faded
 * neighbour word the roll shows on that side.
 */
val Valence.polarityBefore: ValencePolarity?
    get() = ValencePolarity.entries.getOrNull(polarityIndex - 1)

/** The polarity one step to the right, null at the end. */
val Valence.polarityAfter: ValencePolarity?
    get() = ValencePolarity.entries.getOrNull(polarityIndex + 1)

File: apps/web/lib/valence/valence.ts
// The 3×3 valence grid as a pair of ordered axes, plus the index arithmetic the
// two-axis roll walks it with. Port of the `// MARK: - The roll` half of
// `apps/ios/Pebbles/Features/Path/Models/Valence.swift`.
//
// Kept off the components for the same reason iOS kept it off the view: the
// roll's interesting behaviour is stepping, clamping and the ladder order, and
// all three are assertable without a gesture (`valence.test.ts`).
//
// The cell is `(size, polarity)` rather than a nine-case enum because web has
// carried the two axes separately since the first composer — `intensity` and
// `valence` are what the reducer, the draft payload and the wire all speak, and
// a ninth name in the middle would be one more thing to keep in sync.

import {
  POLARITY_BY_VALENCE,
  SIZE_BY_INTENSITY,
  type Intensity,
  type Polarity,
  type Size,
  type Valence,
} from "@/lib/config/pebble-geometry"

/** One of the nine cells: how big the event was, and how it landed. */
export type ValenceCell = { size: Size; polarity: Polarity }

/** Left → right order of the polarity axis. */
export const POLARITY_ORDER: readonly Polarity[] = ["lowlight", "neutral", "highlight"]

/**
 * Top → bottom order of the size axis. Deliberately *not* small → large: the
 * ladder runs big at the top, so dragging up rolls toward smaller events the
 * way scrolling down a list whose big end is at the top does.
 */
export const SIZE_LADDER: readonly Size[] = ["large", "medium", "small"]

export const INTENSITY_BY_SIZE: Record<Size, Intensity> = { small: 1, medium: 2, large: 3 }

export const VALENCE_BY_POLARITY: Record<Polarity, Valence> = {
  lowlight: -1,
  neutral: 0,
  highlight: 1,
}

/** Every cell, small → large and lowlight → highlight within each ring. */
export const ALL_CELLS: readonly ValenceCell[] = (["small", "medium", "large"] as const).flatMap(
  (size) => POLARITY_ORDER.map((polarity) => ({ size, polarity })),
)

export function cellFrom(intensity: Intensity, valence: Valence): ValenceCell {
  return { size: SIZE_BY_INTENSITY[intensity], polarity: POLARITY_BY_VALENCE[valence] }
}

export function cellsEqual(a: ValenceCell, b: ValenceCell): boolean {
  return a.size === b.size && a.polarity === b.polarity
}

/** Cell key, for React lists and for the artwork/outline lookups. */
export function cellKey(cell: ValenceCell): string {
  return `${cell.size}-${cell.polarity}`
}

export function polarityIndex(cell: ValenceCell): number {
  return POLARITY_ORDER.indexOf(cell.polarity)
}

export function sizeIndex(cell: ValenceCell): number {
  return SIZE_LADDER.indexOf(cell.size)
}

function clamp(index: number, length: number): number {
  return index < 0 ? 0 : index > length - 1 ? length - 1 : index
}

/**
 * The cell at the given indices, **clamped** to the grid — the roll stops at
 * the edges rather than wrapping, so a hard swipe cannot loop the user past the
 * end and back to where they started.
 */
export function cellAt(polarity: number, size: number): ValenceCell {
  return {
    polarity: POLARITY_ORDER[clamp(polarity, POLARITY_ORDER.length)],
    size: SIZE_LADDER[clamp(size, SIZE_LADDER.length)],
  }
}

/**
 * The polarity one step to each side, null at the ends. Drives the faded
 * neighbour words the roll bleeds off each screen edge.
 */
export function polarityBefore(cell: ValenceCell): Polarity | null {
  const index = polarityIndex(cell)
  return index > 0 ? POLARITY_ORDER[index - 1] : null
}

export function polarityAfter(cell: ValenceCell): Polarity | null {
  const index = polarityIndex(cell)
  return index < POLARITY_ORDER.length - 1 ? POLARITY_ORDER[index + 1] : null
}
